src/Rules.Framework/Source/RulesSource.cs

```diff
@@ -0,0 +1,183 @@
+namespace Rules.Framework.Source
+{
+    using Rules.Framework.Core;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+
+    internal class RulesSource<TContentType, TConditionType> : IRulesSource<TContentType, TConditionType>
+    {
+        private readonly List<IRulesSourceMiddleware<TContentType, TConditionType>> middlewares;
+        private readonly IRulesDataSource<TContentType, TConditionType> rulesDataSource;
+
+        public RulesSource(
+            IRulesDataSource<TContentType, TConditionType> rulesDataSource,
+            IEnumerable<IRulesSourceMiddleware<TContentType, TConditionType>> middlewares)
+        {
+            this.middlewares = new List<IRulesSourceMiddleware<TContentType, TConditionType>>(middlewares);
+            this.rulesDataSource = rulesDataSource;
+        }
+
+        public async Task AddRuleAsync(AddRuleArgs<TContentType, TConditionType> args)
+        {
+            var queuedMiddlewares = new Queue<IRulesSourceMiddleware<TContentType, TConditionType>>(this.middlewares);
```

Comment:
Why instantiate the queue of middlewares here and not at the constructor? We could avoid having two methods (the public and the internal) for each of the "actions" being done (HandleAddRuleAsync, HandleGetRulesAsync, ...), and have just the public one instead.

Reply:
The reason is that each time we call upon `Dequeue()`, an element is removed from the queue and is never restored again. Next time you call one of the handle methods, the queue will be empty. This was the solution found to circumvent that behavior. I know this has an efficiency penalty on memory and CPU and obliges us to have public and private methods.

I'll be honest with you, I did not dig too deep to find a more efficient alternative (which I believe that we can find), I was more focused on getting the API right for extensibility and left this theme for later. I am open to suggestions to avoid this, I'll take a look at this and get back to you :) shout out if you have any idea about this also 😄

Reply:
Understand the Dequeue part. 👌🏼  Maybe there was not the need for using Queues, recursiveness and delegates (as we're really just iterating over a constant collection of middlewares, and applying each one to the same args), but perhaps I'm not seeing the whole picture. 

Given the "urgency" in make this available and since this is something that can be handled internally, I'd say to proceed and revisit this afterwards.

Reply:
I've experimented with an alternative using .NET's implementation of a `LinkedList`, and it seems that it is better now 👌 

In a few words, explaining the change, now during `RulesSource` initialization we use a `LinkedList` to represent the ordered middleware collection. This allows us to build a middleware "pipeline" from the last middleware to the first, and more importantly, it is only built once and re-used every time one of the `RulesSource` are called upon.

Glad you opened this discussion, it is much better now! 😄 Can you retake a look and tell me what you think?

Reply:
Cool, it definitely looks better now! 💪🏼  

My additional doubts have to do with understanding the whole structure and the work of the delegates in it.

If there's more than one middleware, each RulesSource method will iterate over the middlewares and chain the actions.. does that mean that the instruction `await rulesDataSource.AddRuleAsync(args.Rule)` will only be called once? 

Meaning that what this structure does is to add layers before and after the core instruction, but it only performs the core action (e.g. `AddRuleAsync()` ) once? 
```
{Middleware1CodeBeforeAction}
{Middleware#2CodeBeforeAction}
CoreAction
{Middleware2CodeAfterAction}
{Middleware#1CodeAfterAction}
```

Took me a bit to have a grasp on that, but if that's the case I now think I got it. It would be hard / less elegant to accomplish this without delegates, right?

I'm asking this because they obviously add complexity to the code, but if I understood correctly, in this case it seems to be justified.

Reply:
That's right, `AddRuleAsync(...)` is called only once 👌 the main idea is to control when/if to call the core action as you described. The effect is similar to using the Decorator pattern.

We could be following another approach, such as using a sort of interceptor implementation (before the core action method and after the core action method) but it would take away from you the flexibility of deciding not to call the core action. Maybe with `AddRuleAsync(...)` it might be difficult to imagine when we would not call the core action, but let's take another example: the `GetRulesAsync(...)`. If you have an implementation for a cache middleware, you can decide not to call the core action `GetRulesAsync(...)` if you have the response cached already. This point here justifies the additional complexity. It could also be accomplished with a Decorator pattern as mentioned above. In this case, I opted for this one for similarity with the ASP.NET pipeline, I thought it would be a concept easier for people to understand.

Reply:
Understood! 👌🏼 Thanks for the explanation! 🙂